flycheck.el

```diff
@@ -7744,7 +7744,12 @@ Otherwise return the previously used cache directory."
 Look up the directory hierarchy from FILE for a directory
 containing a file that matches REGEXP."
   (locate-dominating-file
-   file
+   ;; We must not pass regular file here because
+   ;; `locate-dominating-file' will just pass it into its second
+   ;; lambda argument . But our lambda expects only directories.
+   (if (file-directory-p file)
+       file
+     (file-name-directory file))
```

Comment:
>  I still don't understand all the reasons why shouldn't flycheck--locate-dominating-file-matching be restricted to requiring directories?

Coherence with how `locate-dominating-file` was already used, mostly.

But if we actually have been using `locate-dominating-file` wrong all this time, then there is no reason to held onto it.  (Note: we should probably hunt these down in the code now)

Looks like you had the right idea all along @sergv  👍  Sorry for the back-and-forth.

Please restore the `DIRECTORY` as first argument to the new flycheck function (we don't need the comment anymore).

Please also squash you commits as one before merging.